### DISCUSSION TITLE
Question: Get actor name inside interaction

### DISCUSSION THREAD

**Discussion by muRn** [serenity-js/serenity-js#1818] [How to... ?] [answered]
Hi, Jan! Hi, everyone!

I'm using serenity js with rest/axios and I have few different actors used across different test cases. I'd like to send actor's name as `x-user-id` header in each of my requests. For now I've only figured out this way of doing it:
```ts
const actor = actorInTheSpotlight();
actor.attemptsTo(
    Send.a(
        PostRequest
        .to('/some/endpoint')
        .with({
            some: 'payload'
        })
        .using({
            headers: {
                'x-user-id': actor.name
            }
        })
    )
);
```

Is there any way to extend/modify `Send` interaction so that the customized version automatically sends actor's name in a header without me repeating this "headers setup" piece of code in each and every test case? Because as I understand interaction is never executed on it's own, there is always some actor executing it, so I thought there should be a way of obtaining details about this "executor". Or is this something against Screenplay pattern and interaction should never care about who executes it?

Thanks!

**Comment by jan-molak** [comment, marked as answer]
Hi @muRn!

Sure, that's a perfectly reasonable approach. You can configure default headers when you give the actor the ability to [`CallAnApi`](https://serenity-js.org/api/rest/class/CallAnApi/), for example:

```typescript
import { actorCalled } from '@serenity-js/core'
import { CallAnApi, GetRequest, LastResponse, Send } from '@serenity-js/rest'
import { Ensure, equals } from '@serenity-js/assertions'

import * as axios from 'axios'

const axiosInstance = axios.create({
  timeout: 5 * 1000,
  headers: {
    'X-Custom-Api-Key': 'secret-key',
  },
});

await actorCalled('Apisitt')
  .whoCan(
    CallAnApi.using(axiosInstance),
  )
  .attemptsTo(
    Send.a(GetRequest.to('/users/2')),
    Ensure.that(LastResponse.status(), equals(200)),
  )
```

To do that automatically for every actor, set up your [`Cast`](https://serenity-js.org/api/core/class/Cast/) of actors to [automatically configure](https://serenity-js.org/api/core/class/SerenityConfig/) the ability to `CallAnApi` when the actor is instantiated. For example with [Mocha](https://serenity-js.org/api/mocha/) you can do it like this:

```typescript
import { Cast, configure } from '@serenity-js/core'
import { CallAnApi } from '@serenity-js/rest'

import * as axios from 'axios'

configure({
  actors: Cast.where(actor => {
    const axiosInstance = axios.create({
      headers: {
        'x-user-id': actor.name,
      },
    });

    return actor.whoCan(
      CallAnApi.at('http://api.example.org'),
      // other abilities
    );
  )
})
```
Then in your test:
```typescript
await actor.attemptsTo(
    Send.a(
        PostRequest
        .to('/some/endpoint')
        .with({
            some: 'payload'
        })
    )
);
```

**Reply by muRn** [reply]
Works like a charm! Thanks a lot :)
